src/Checkout/ExpressCheckout/ExpressCheckoutSubscriber.php

```diff
@@ -227,6 +217,39 @@ public function addPayerIdToCustomer(DataMappingEvent $event): void
         $event->setOutput($output);
     }
 
+    /**
+     * If double opt-in is enabled, we need to patch the customer to disable it again.
+     * Otherwise, we will not get a logged-in customer and an email will be sent, requiring confirmation.
+     * We don't need this confirmation as PayPal is our trusted source.
+     */
+    public function onCustomerWritten(EntityWrittenEvent $event): void
+    {
+        if (!$event->getContext()->hasExtension(ExpressCustomerService::EXPRESS_CHECKOUT_ACTIVE)) {
+            return;
+        }
+
+        // as we're in express checkout, there will be only one write result
+        foreach ($event->getWriteResults() as $result) {
+            $id = $result->getProperty('id');
+            $salesChannelId = $result->getProperty('salesChannelId');
+            $expressId = $result->getProperty('customFields')[ExpressCustomerService::EXPRESS_PAYER_ID] ?? null;
+            $guest = (bool) $result->getProperty('guest');
+
+            // double-checking if we should patch
+            if (!$id || !$salesChannelId || !$expressId || !$guest) {
+                continue;
+            }
+
+            if ($this->systemConfigService->getBool('core.loginRegistration.doubleOptInGuestOrder', $salesChannelId)) {
+                $this->customerRepository->update([[
+                    'id' => $id,
+                    'doubleOptInRegistration' => false,
+                    'doubleOptInEmailSentDate' => null,
+                ]], $event->getContext());
```

Comment:
couldn't you get into an infinite loop by writing the customer again? 🤔

Reply:
As the payload doesn't provide the information I need (salesChannelId, guest, expressId) it is skipped. If you prefer, I could be more explicit by manipulating the context and remove the express checkout flag

Reply:
I think this is fine.

However I'm not sure about the `!$id` check. Won't the payload have `$id` every time?

Reply:
I assume this is a PHPstan check?

Reply:
Of course, but we're not patching based on only having the `$id` provided :thinking:
I changed the context to not include the express checkout extension, just in case. Should be more future prove this way